### DISCUSSION TITLE
No Windows console found. Are you running cmd.exe?

### DISCUSSION THREAD

**Discussion by swanknight** [lncrawl/lightnovel-crawler#1270] [General]
I'm trying to create a GUI for **lncrawl** (using .NET 5) but I cam across an odd issue, when I try to run **lncrawl** from my UI I keep getting an error: _"No Windows console found. Are you running cmd.exe?"_

These are the parameters I used for my test:

`lncrawl -s https://novelfull.com/i-alone-level-up.html -o "D:\" --all -f --format epub --single`

**Comment by dipu-bd** [comment]
Try like this:

```C#
using System.Diagnostics;

Process process = new Process();
ProcessStartInfo startInfo = new ProcessStartInfo();
startInfo.WindowStyle = ProcessWindowStyle.Hidden;
startInfo.FileName = "cmd.exe";
startInfo.Arguments = "/C lncrawl -s https://novelfull.com/i-alone-level-up.html -o "D:\" --all -f --format epub --single";
process.StartInfo = startInfo;
process.Start();
```

**Comment by swanknight** [comment]
> Try like this:
> 
> ```cs
> using System.Diagnostics;
> 
> Process process = new Process();
> ProcessStartInfo startInfo = new ProcessStartInfo();
> startInfo.WindowStyle = ProcessWindowStyle.Hidden;
> startInfo.FileName = "cmd.exe";
> startInfo.Arguments = "/C lncrawl -s https://novelfull.com/i-alone-level-up.html -o "D:\" --all -f --format epub --single";
> process.StartInfo = startInfo;
> process.Start();
> ```

It doesn't work. I need to redirect the output to my UI (RichTextox) and it doesn't work if I try to use "cmd".
I've made plenty of GUI applications before but I never needed to run "cmd", so something is odd with **lncrawl**.

**Comment by dipu-bd** [comment]
Add `--suppress` argument. it will stop prompting any command line inputs. e.g.: `lncrawl -s https://novelfull.com/i-alone-level-up.html -o "D:\" --all -f --format epub --single --suppress`

**Comment by swanknight** [comment]
> --suppress

That did fix the error but for some reason, it's **not** creating the epub file. I can see that it created the json folder and downloaded the content but nothing happens in the end. This is the output I get in my "console":

```
================================================================================
                         [#] Lightnovel Crawler v2.28.11 
                  https://github.com/dipu-bd/lightnovel-crawler 
--------------------------------------------------------------------------------
 ! Input is suppressed
--------------------------------------------------------------------------------
Namespace(add_source_url=False, all=True, bot=None, chapters=None, crawler=[], extra={}, filename=None, filename_only=False, first=None, force=True, ignore=False, last=None, list_sources=False, log=None, login=None, multi=False, novel_page='https://novelfull.com/i-alone-level-up.html', output_formats=['epub'], output_path='D:\\', page=None, query=None, range=None, shard_count=1, shard_id=0, single=True, suppress=True, volumes=None)

-> Press  Ctrl + C  to exit

Retrieving novel info...
https://novelfull.com/i-alone-level-up.html
NOVEL: I Alone Level-Up
3 volumes and 270 chapters found
```

**Comment by dipu-bd** [comment]
it is showing `output_formats=['epub']` as argument. that means the problem is somewhere else. try running it again with `-lll` option. it will print more details log

**Comment by swanknight** [comment]
> 

Adding `-lll` doesn't seem to display anything new, and it doesn't download anything after adding it.
Also, it's strange that even though that without `-lll` it does download the chapters and creates the json folder, I see no progress in the log.

```
================================================================================
                         [#] Lightnovel Crawler v2.28.11 
                  https://github.com/dipu-bd/lightnovel-crawler 
--------------------------------------------------------------------------------
                          << LOG LEVEL: DEBUG
--------------------------------------------------------------------------------
 ! Input is suppressed
--------------------------------------------------------------------------------
Namespace(add_source_url=False, all=True, bot=None, chapters=None, crawler=[], extra={}, filename=None, filename_only=False, first=None, force=True, ignore=False, last=None, list_sources=False, log=3, login=None, multi=False, novel_page='https://novelfull.com/i-alone-level-up.html', output_formats=['epub'], output_path='D:\\', page=None, query=None, range=None, shard_count=1, shard_id=0, single=True, suppress=True, volumes=None)

-> Press  Ctrl + C  to exit

Retrieving novel info...
https://novelfull.com/i-alone-level-up.html
NOVEL: I Alone Level-Up
3 volumes and 270 chapters found
```

**Comment by dipu-bd** [comment]
Without any error log, I don't know how to solve the issue.

There was an old issue for command line encoding. #927 #882 #892 

Check if this is the cause.

**Comment by swanknight** [comment]
> Without any error log, I don't know how to solve the issue.
> 
> There was an old issue for command line encoding. #927 #882 #892
> 
> Check if this is the cause.

Unfortunately the `-lll` doesn't display anything and as I said, it even refuses to even download when used.
The strange thing is when I ran lncrawl separately through cmd using the exact same parameters it's working fine.

**_Update_**: _funny thing, when I used `i` instead of `f` it worked...sometimes. Yep, using the exact same parameters it worked sometimes and failed sometimes. This is too bizzare._

**Comment by dipu-bd** [comment]
Any update on this? Can you check the latest version?

**Reply by swanknight** [reply]
> Any update on this? Can you check the latest version?

Hi again, I finally figured it out! I use `RedirectStandardOutput` to get *lncrawl*'s output into my form, but for some strange reason the progress part of the output is not actually treated as "Output", instead it's treated as an "Error". 

I have no idea why this happens, but the workaround is to also use `RedirectStandardError` and combine it with `RedirectStandardOutput` into a single event to get the full output. And it worked!

That's the good news. The **bad news** is that I ran into another issue. *lncrawl* doesn't close automatically after it finishes, instead it waits for input from the user (press Ctrl + C?) which cause the process to remain in the memory even after it completes the download, which in turn causes many issues in my code.

I imagine a somewhat simple solution is that you add a new argument _(ex: --auto-close)_ that automatically closes the console window once it finishes without requiring a user command. What do you think? Or is there another way to achieve this?

**Reply by dipu-bd** [reply]
It is a windows only feature. Window's exe builds closes automatically after it is done. But some users did not want that to happen.  Since you can not even see what the error is if it closes automatically after it is done.

I suggest modifying the source as you need, and building your own exe package. Just run `python setup.py package` to make an exe.

Do these modifications to `lncrawl/core/__init__.py` to get what you want:

- Remove [line 88 to 94](https://github.com/dipu-bd/lightnovel-crawler/blob/f6aa48ccc955075796255922a3eb788e71f500df/lncrawl/core/__init__.py#L88-L94) to hide Ctrl + C at the end.
- Insert `os.environ['LOG_LEVEL'] = 'DEBUG'` after [line 24](https://github.com/dipu-bd/lightnovel-crawler/blob/f6aa48ccc955075796255922a3eb788e71f500df/lncrawl/core/__init__.py#L24) if you want to hide the progressbars

**Reply by swanknight** [reply]
> It is a windows only feature. Window's exe builds closes automatically after it is done. But some users did not want that to happen. Since you can not even see what the error is if it closes automatically after it is done.
> 
> I suggest modifying the source as you need, and building your own exe package. Just run `python setup.py package` to make an exe.
> 
> Do these modifications to `lncrawl/core/__init__.py` to get what you want:
> 
> * Remove [line 88 to 94](https://github.com/dipu-bd/lightnovel-crawler/blob/f6aa48ccc955075796255922a3eb788e71f500df/lncrawl/core/__init__.py#L88-L94) to hide Ctrl + C at the end.
> * Insert `os.environ['LOG_LEVEL'] = 'DEBUG'` after [line 24](https://github.com/dipu-bd/lightnovel-crawler/blob/f6aa48ccc955075796255922a3eb788e71f500df/lncrawl/core/__init__.py#L24) if you want to hide the progressbars

I understand that not everyone would want that, hence my suggestion of it being an _optional_ argument/parameter.
As for modifying the source myself, I'm afraid I know nothing about Python and even if I did, it wouldn't be a good idea. 
I'm trying to create a generic GUI for _lncrawl_ which would automatically download the latest version from GitHub. 
I don't wanna lock it to only one custom version, it will lead to problems down the line for sure.

**Reply by dipu-bd** [reply]
I have added a command `--close-directly` to suppress the input. It will be available from the next release.

**Reply by swanknight** [reply]
> I have added a command `--close-directly` to suppress the input. It will be available from the next release.

That's awesome, thanks.
